shell/platform/darwin/macos/framework/Source/FlutterKeyEmbedderHandler.mm

```diff
@@ -0,0 +1,605 @@
+      codeUnit = (codeUnit << 16) | secondCode;
+    }
+    if (codeUnit < 256) {
+      if (isupper(codeUnit)) {
+        return tolower(codeUnit);
+      }
+      return codeUnit;
+    }
+    return KeyOfPlane(codeUnit, kUnicodePlane);
+  }
+
+  // Control keys like "backspace" and movement keys like arrow keys don't have
+  // a printable representation, but are present on the physical keyboard. Since
+  // there is no logical keycode map for macOS (macOS uses the keycode to
+  // reference physical keys), a LogicalKeyboardKey is created with the physical
+  // key's HID usage and debugName. This avoids duplicating the physical key
+  // map.
+  if (physicalKey != 0) {
+    return KeyOfPlane(physicalKey, kHidPlane);
+  }
+
+  // This is a non-printable key that is unrecognized, so a new code is minted
+  // with the autogenerated bit set.
+  return KeyOfPlane(event.keyCode, kMacosPlane | kAutogeneratedMask);
+}
+
+/**
+ * Converts NSEvent.timestamp to the timestamp for Flutter.
+ */
+static double GetFlutterTimestampFrom(NSTimeInterval timestamp) {
+  // Timestamp in microseconds. The event.timestamp is in seconds with sub-ms precision.
+  return timestamp * 1000000.0;
+}
+
+/**
+ * Compute |modifierFlagOfInterestMask| out of |keyCodeToModifierFlag|.
+ *
+ * This equals to the bitwise-or of all values of |keyCodeToModifierFlag| as well as
+ * NSEventModifierFlagCapsLock.
+ */
+static NSUInteger computeModifierFlagOfInterestMask() {
+  __block NSUInteger modifierFlagOfInterestMask = NSEventModifierFlagCapsLock;
+  [keyCodeToModifierFlag
+      enumerateKeysAndObjectsUsingBlock:^(NSNumber* keyCode, NSNumber* flag, BOOL* stop) {
+        modifierFlagOfInterestMask = modifierFlagOfInterestMask | [flag unsignedLongValue];
+      }];
+  return modifierFlagOfInterestMask;
+}
+
+/**
+ * The handler sent to the embedder's SendKeyEvent.
+ *
+ * The |user_data| should actually be a |FlutterKeyPendingResponse| that has
+ * been retained once with |__bridge_retained|. This function calls
+ * |FlutterKeyPendingResponse.handler|'s |handleResponse| using |handled|.
+ */
+void HandleResponse(bool handled, void* user_data);
+
+/*
+ * Converts NSEvent.characters to a C-string for FlutterKeyEvent.
+ */
+const char* getEventString(NSString* characters) {
+  if ([characters length] == 0) {
+    return nullptr;
+  }
+  unichar utf16Code = [characters characterAtIndex:0];
+  if (utf16Code >= 0xf700 && utf16Code <= 0xf7ff) {
+    // Some function keys are assigned characters with codepoints from the
+    // private use area (see
+    // https://developer.apple.com/documentation/appkit/1535851-function-key_unicodes?language=objc).
+    // These characters are filtered out since they're unprintable.
+    //
+    // Although the documentation claims to reserve 0xF700-0xF8FF, only up to 0xF747
+    // is actually used. Here we choose to filter out 0xF700-0xF7FF section.
+    // The reason for keeping the 0xF800-0xF8FF section is because 0xF8FF is
```

Comment:
I feel like they're using the end of the region (0xF8FF and backwards) for printable characters, and the front of the region of unprintable keys. I'd like to keep the printable characters if there are more that I haven't tested or in case they add more.